students/mattmaeda/session08/sparse_array.py

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
+"""
+Sparse array implementation
+"""
+
+class SparseArray(object):
+    def __init__(self, number_array):
+        self.__length = len(number_array)
+        self.__array = {}
+
+        for index, value in enumerate(number_array):
+            if value != 0:
+                self.__array[index] = value
+
+
+    def construct_list(self):
+        return [self.__array[i] if i in self.__array else 0 for i in
+                range(self.__length)]
+
+
+    def __str__(self):
+        return "[{}]".format(",".join([str(i) for i in self.construct_list()]))
+
+
+    def __repr__(self):
+        return "[{}]".format(",".join([str(i) for i in self.construct_list()]))
+
+
+    def __len__(self):
+        return self.__length
+
+
+    def __getitem__(self, index):
+        if index > self.__length - 1:
+            raise IndexError
+        elif index not in self.__array:
+            return 0
+        return self.__array[index]
+
+
+    def __delitem__(self, index):
+        if index > self.__length - 1:
+            raise IndexError
+        else:
+            for i in range(index, (self.__length - 1)):
+                if (i+1) in self.__array:
+                    self.__array[i] = self.__array[i+1]
+
+            self.__array.pop((self.__length - 1), None)
+            self.__length -= 1
+
+
+    def __iter__(self):
+        self.__index = 0
+        return self
+
+
+    def __next__(self):
+        self.__index += 1
+
+        if self.__length >= self.__index:
+            if (self.__index - 1) in self.__array:
+                return self.__array[(self.__index - 1)]
+            return 0
+
+        else:
+            raise StopIteration
+
```

Comment:
This is subtle, but your iterator does not work quite like other python iterators:
```
In [6]: sparse
Out[6]: [1,0,1,3,9,9]

In [7]: test = iter(sparse)

In [8]: next(test)
Out[8]: 1

In [9]: next(test)
Out[9]: 0

In [10]: next(test)
Out[10]: 1

In [11]: for i in test:
    ...:     print(i)
    ...:
1
0
1
3
9
9

In [12]: new = [1, 2, 3, 4, 5]

In [13]: test2 = iter(new)

In [14]: next(test2)
Out[14]: 1

In [15]: next(test2)
Out[15]: 2

In [16]: for i in test2:
    ...:     print(i)
    ...:
3
4
5
```